src/crypto/bcrypt.cr

```diff
@@ -1,153 +1,119 @@
 require "secure_random"
-require "./blowfish"
 require "./subtle"
-require "./bcrypt/base64"
 
-module Crypto::Bcrypt
-  extend self
+# Pure Crystal implementation of the Bcrypt algorythm by Niels Provos and David
+# Mazières, as [presented at USENIX in
+# 1999](https://www.usenix.org/legacy/events/usenix99/provos/provos_html/index.html).
+#
+# Refer to `Crypto::Bcrypt::Password` for a higher level interface.
+#
+# About the Cost
+#
+# Bcrypt, like the PBKDF2 or scrypt ciphers, are designed to be slow, so
+# generating rainbow tables or cracking passwords is nearly impossible. Yet,
+# computers are always getting faster and faster, so the actual cost must be
+# incremented every once in a while.
+# Always use the maximum cost that is tolerable, performance wise, for your
+# application. Be sure to test and select this based on your server, not your
+# home computer.
+#
+# This implementation of Bcrypt is currently 50% slower than pure C solutions,
+# so keep this in mind when selecting your cost. It may be wise to test with
+# Ruby's bcrypt gem which is a binding to OpenBSD's implementation.
+#
+# Last but not least: beware of denial of services! Always protect your
+# application using an external strategy (eg: rate limiting), otherwise
+# endpoints that verifies bcrypt hashes will be an easy target.
+class Crypto::Bcrypt
+  class Error < Exception
+  end
 
-  MIN_COST = 4
-  MAX_COST = 63
   DEFAULT_COST = 10
```

Comment:
A default cost of 10 may be a very bad default now: it takes 100ms (60ms in C) to hash a password on my laptop. Maybe a cost of 12 (400ms, but 240ms in C) would be a better default.